Add prompting mechanism for config values and add global options: --config, --server, --username, --password, --debug

msabramo opened the pull request:
This allows not keeping passwords and such in the config file.

E.g.: In `/etc/aptly-cli.conf`:

``` yaml

---
:proto: http
:server: 10.3.0.46
:port: 8082
:debug: false
:username: ${PROMPT}
:password: ${PROMPT_PASSWORD}
```

The tool will prompt for the specified values, where `${PROMPT}` results in a regular prompt and `${PROMPT_PASSWORD}` results in a password prompt where the input is replaced by asterisks -- e.g.:

```
$ aptly-cli version
Enter a value for username:
zane
Enter a value for password:
***************
{"Version"=>"0.9.7"}
```

This also eliminates a bunch of duplicated code in all of the command classes by having them inherit from a common base class. This was actually essential because otherwise, every required file would trigger a config load and thus prompting for any needed values over and over again.

The `--config` (`-c`) option allows specifying an alternative config file -- e.g.:

```
$ aptly-cli -c ~/.config/aptly-cli/aptly-cli.conf repo_list
```

and the `--server`, `--username`, and `--password` options allow specifying those things on the command-line and not even requiring a config file:

```
$ aptly-cli --server 10.3.0.46 --username marca --password '${PROMPT_PASSWORD}' repo_list
```

Note that you can use `${PROMPT}` and `${PROMPT_PASSWORD}` in the values of these options, just as you can in a config file.

msabramo committed:
Add prompting mechanism for config values

This allows not keeping passwords and such in the config file.

E.g.: In `/etc/aptly-cli.conf`:

```yaml
---
:proto: http
:server: 10.3.0.46
:port: 8082
:debug: false
:username: ${PROMPT}
:password: ${PROMPT_PASSWORD}
```

The tool will prompt for the specified values -- e.g.:

```
$ aptly-cli version
Enter a value for username:
zane
Enter a value for password:
***************
{"Version"=>"0.9.7"}
```

This also eliminates a bunch of duplicated code in all of the command
classes by having them inherit from a common base class. This was
actually essential because otherwise, every required file would trigger
a config load and thus prompting for any needed values over and over
again.

msabramo committed:
Add --config global option

This lets you specify a different config file. E.g.:

    aptly-cli -c ~/.config/aptly-cli/aptly-cli.conf repo_list

msabramo committed:
--server, --username, --password global options

This required a fairly big refactor, because I had to move stuff that
processes `${PROMPT}` and `${PROMPT_PASSWORD}` from `aptly_load.rb` to
`aptly_command.rb`, so that it could be used in the global options as
well as in the configuration file.

msabramo committed:
Add --debug global option

sepulworld commented:
This is great! Looks good first pass on my phone. I'll take a closer look this weekend or Monday for sure.

msabramo committed:
Use `keyring` gem when `${KEYRING}` specified

Optionally use the [`keyring` gem](https://github.com/jheiss/keyring)
when `${KEYRING}` is specified as a value for an option or config
setting.

Note that I require `keyring` at the point of use, so that it is an
optional dependency.

sepulworld commented:
Ok, this looks good to merge to me!  Should we add your pull request notes to the README.md for docs too?  I can merge this and https://github.com/sepulworld/aptly_cli/pull/101 then release gem 0.3.0.  Ill tag the release with notes on all of the changed add.  Thanks again, this is a great security and functionality addition!

sepulworld commented:
https://github.com/sepulworld/aptly_cli/releases/tag/0.3.0

sepulworld commented:
https://github.com/sepulworld/aptly_cli/pull/102 for adding to README.md this feature

msabramo commented:
https://github.com/sepulworld/aptly_cli/pull/105 documents the keyring stuff in README.md
